onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -702,7 +713,7 @@ class PlannerImpl {
               }
             }
           }
-        } else if (IsNonTensor(*node_output)) {
+        } else if (IsNonTensor(*node_output) || external_outputs) {
```

Comment:
Hi @iK1D , thanks for the change. Have you verified the outputs are actually not allocated? It seems with this change Yield Op's output tensors will be marked as "kAllocate" which will make execution frame [allocate](https://github.com/microsoft/onnxruntime/blob/5b6753ce278166bf1f979eee1643a0bec6696c40/onnxruntime/core/framework/execution_frame.cc#L554) the tensor(s). My expectation was we will introduce a type of allocation akin to "[kPreExisting](https://github.com/microsoft/onnxruntime/blob/5b6753ce278166bf1f979eee1643a0bec6696c40/include/onnxruntime/core/framework/alloc_kind.h#L28)" that we use for inputs, or *may be* we might be able to re-use it.

Reply:
Sherlock tells me there are two implicit assumptions made 1) mem_planning will be turned off 2) GetOutputs will not called, really none of these are enforced. kAllocate in the codebase means allocate when the kernel requests that output.

Reply:
indeed... reuse kPreExisting looks more appropriate here

Reply:
@SherlockNoMad Initially missed this question that you asked but in the event memory pattern is turned on and even if you don't access the output tensor the memory pattern generation will factor these tensors in the static buffer because they are kAllocate. The pattern generation logic ONLY accounts for kAllocate tensors.

Reply:
The 'allocate' code is called only from GetOrCreateNodeOutputMLValue, and since we don't call it for YieldOp outputs, it's indeed not allocated when mem_planning is turned off.
I actually also tried both kAllocate and kPreExisting for mem_planning is turned on, and it doesn't make any difference for the BERT. When mem_planning is turning on, we will calculate the peak memory and allocate the buffer, not all kAllocate outputs will contribute to the peak, maybe for the BERT case here, the peak is not at this place during the calculation. But yes, I think maybe kPreExisting is more reasonable, will make the change.

Reply:
@iK1D If an output tensor is marked as kAllocate then it will be factored in static buffer when memory pattern is turned on and can contribute to peak memory, please refer [this ](https://github.com/microsoft/onnxruntime/blob/40fa40f3ce7419a093467c70dc8ff8fcfed9335f/onnxruntime/core/framework/session_state.cc#L495)line.